### ISSUE TITLE
FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory

### ISSUE THREAD

**Issue by NextFire** [microsoft/pyright#5540]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

If you have a question about typing or a behavior that you’re seeing in Pyright (as opposed to a bug report or enhancement request), consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Describe the bug**

Since version 1.1.308

```
> npx pyright@1.1.318 --verbose foobar.py
Search paths for /Users/namvu/Developer/test
  /Users/namvu/.npm/_npx/f63f8f904f6b995a/node_modules/pyright/dist/typeshed-fallback/stdlib
  /Users/namvu/Developer/test
  /Users/namvu/Developer/test/typings
  /Users/namvu/.npm/_npx/f63f8f904f6b995a/node_modules/pyright/dist/typeshed-fallback/stubs/...
  /usr/local/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11
  /usr/local/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/lib-dynload
  /Users/namvu/Developer/test/.venv/lib/python3.11/site-packages
Found 1 source file

<--- Last few GCs --->

[21990:0x7f930d100000]    23786 ms: Scavenge 2039.3 (2077.1) -> 2038.1 (2077.9) MB, 5.37 / 0.00 ms  (average mu = 0.257, current mu = 0.245) allocation failure; 
[21990:0x7f930d100000]    23791 ms: Scavenge 2040.1 (2077.9) -> 2038.8 (2078.4) MB, 3.12 / 0.00 ms  (average mu = 0.257, current mu = 0.245) allocation failure; 
[21990:0x7f930d100000]    23807 ms: Scavenge 2040.6 (2078.4) -> 2039.4 (2083.6) MB, 4.47 / 0.00 ms  (average mu = 0.257, current mu = 0.245) allocation failure; 


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x10bfb760f node::Abort() [/usr/local/Cellar/node/20.4.0/bin/node]
 2: 0x10bfb8cde node::ModifyCodeGenerationFromStrings(v8::Local<v8::Context>, v8::Local<v8::Value>, bool) [/usr/local/Cellar/node/20.4.0/bin/node]
 3: 0x10c12cc35 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, v8::OOMDetails const&) [/usr/local/Cellar/node/20.4.0/bin/node]
 4: 0x10c12cbcb v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, v8::OOMDetails const&) [/usr/local/Cellar/node/20.4.0/bin/node]
 5: 0x10c2ab373 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/usr/local/Cellar/node/20.4.0/bin/node]
 6: 0x10c2adeee v8::internal::Heap::ComputeMutatorUtilization(char const*, double, double) [/usr/local/Cellar/node/20.4.0/bin/node]
 7: 0x10c2ac01b v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::internal::GarbageCollectionReason, char const*) [/usr/local/Cellar/node/20.4.0/bin/node]
 8: 0x10c2a9f6b v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/Cellar/node/20.4.0/bin/node]
 9: 0x10c2a9bfd v8::internal::Heap::HandleGCRequest() [/usr/local/Cellar/node/20.4.0/bin/node]
10: 0x10c257192 v8::internal::StackGuard::HandleInterrupts() [/usr/local/Cellar/node/20.4.0/bin/node]
11: 0x10c5a674c v8::internal::Runtime_StackGuardWithGap(int, unsigned long*, v8::internal::Isolate*) [/usr/local/Cellar/node/20.4.0/bin/node]
12: 0x10bde2af6 Builtins_CEntry_Return1_ArgvOnStack_NoBuiltinExit [/usr/local/Cellar/node/20.4.0/bin/node]
fish: Job 1, 'npx pyright@1.1.318 --verbose f…' terminated by signal SIGABRT (Abort)
```

**To Reproduce**

Install `aiohttp` and `backoff` on Python 3.11 then run pyright on the code snippet below.

**Expected behavior**

```
> npx pyright@1.1.307 --verbose foobar.py
Search paths for /Users/namvu/Developer/test
  /Users/namvu/.npm/_npx/8a0d47663decad1d/node_modules/pyright/dist/typeshed-fallback/stdlib
  /Users/namvu/Developer/test
  /Users/namvu/Developer/test/typings
  /Users/namvu/.npm/_npx/8a0d47663decad1d/node_modules/pyright/dist/typeshed-fallback/stubs/...
  /usr/local/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11
  /usr/local/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/lib-dynload
  /Users/namvu/Developer/test/.venv/lib/python3.11/site-packages
Found 1 source file
pyright 1.1.307
0 errors, 0 warnings, 0 informations 
Completed in 0.878sec
```

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or imports all referenced symbols.

```py
import aiohttp
import backoff

conn_backoff = backoff.on_exception(
    backoff.expo,
    (aiohttp.ClientConnectorError, aiohttp.ClientConnectionError,
     aiohttp.ContentTypeError),
    max_time=600,
)


def giveup(exception):
    return 400 <= exception.status < 500


response_backoff = backoff.on_exception(backoff.expo,
                                        aiohttp.ClientResponseError,
                                        max_time=600,
                                        giveup=giveup)

timeout_backoff = backoff.on_exception(
    backoff.expo,
    aiohttp.ServerTimeoutError,
    max_time=300,
    max_tries=5
)

default_backoff = timeout_backoff(conn_backoff(response_backoff))
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

```requirements
aiohttp==3.8.4
aiosignal==1.3.1
async-timeout==4.0.2
attrs==23.1.0
backoff==2.2.1
charset-normalizer==3.2.0
frozenlist==1.4.0
idna==3.4
multidict==6.0.4
yarl==1.9.2
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, or the command-line tool? Which version?

CLI, node v20.4.0

VS Code Python/Pylance extension syntax highlighting also stopped working, it keeps saying "1 file to analyze" at the bottom left.

**Additional context**

**Comment by erictraut** [comment]
Thanks for the bug report and the clear repro steps. I'm able to repro the problem, and I'll investigate further.

**Comment by erictraut** [comment]
Here's a minimal repro:

```python
from typing import Callable, TypeVar

_T = TypeVar("_T")
def func0() -> Callable[[_T], _T]: ...

f1 = func0()
func1(func1)
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.319, which I just published. It will also be included in a future release of pylance.
